homeassistant/helpers/script.py

```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
         if not unsub:
             return
 
-        if not self._stop.is_set():
-            self._changed()
+        self._changed()
         try:
-            timeout = self._action[CONF_TIMEOUT].total_seconds()
+            delay = self._action[CONF_TIMEOUT].total_seconds()
         except KeyError:
-            timeout = None
+            delay = None
         done = asyncio.Event()
         try:
-            await asyncio.wait_for(
-                asyncio.wait(
+            async with timeout(delay):
+                _, pending = await asyncio.wait(
                     {self._stop.wait(), done.wait()},
                     return_when=asyncio.FIRST_COMPLETED,
-                ),
-                timeout,
-            )
+                )
+            for pending_task in pending:
+                pending_task.cancel()
         except asyncio.TimeoutError:
             if not self._action.get(CONF_CONTINUE, True):
                 self._log(_TIMEOUT_MSG)
                 raise _StopScript
         finally:
             unsub()
 
+    async def _async_call_service_step(self):
+        """Call the service specified in the action."""
+        domain, service, service_data = self._prep_call_service_step()
+        blocking = service_data.get(ATTR_BLOCKING, True)
+
+        # If this might start a fully blocking script call then disable the call
+        # timeout. Otherwise use the normal service call limit.
+        if domain == "script" and (
+            service == SERVICE_TURN_ON and blocking or service != SERVICE_TURN_OFF
+        ):
+            limit = None
+        else:
+            limit = SERVICE_CALL_LIMIT
+
+        coro = self._hass.services.async_call(
+            domain,
+            service,
+            service_data,
+            blocking=True,
+            context=self._context,
+            limit=limit,
+        )
 
-class _BackgroundScriptRun(_ScriptRun):
-    """Manage background Script sequence run."""
+        if limit is None:
+            # No call limit (i.e., potentially starting a fully blocking script call) so
+            # watch for a stop request and, if received, propagate it to the called
+            # scripts by turning them off.
+            task = self._hass.async_create_task(coro)
+            done, pending = await asyncio.wait(
+                {self._stop.wait(), task}, return_when=asyncio.FIRST_COMPLETED,
+            )
+            for pending_task in pending:
+                if pending_task == task:
+                    # The scripts did not all finish. Stop them by turning them off.
```

Comment:
We should be able to cancel our task and let asyncio do the work. We should not add custom logic because we can never support everything. Since scripts are shielded, they will not just stop half-way.

Reply:
The intent is to stop scripts gracefully (i.e., between steps, or to stop the current step if it's a `delay`, `wait_template` or a call to another script.) And, yes, they are shielded, and if the task that calls `Script.asyn_run` is cancelled it will handle it and stop gracefully.

But this is not in the script or automation integration, so it doesn't have direct access to the corresponding entities. More to the point, it doesn't even necessarily have direct access to the tasks that called the `async_run` method of the potentially several `Script` objects. So instead it's using the service `script.turn_off` to tell those scripts / automation action sequences to stop. We are at the service call level here, so it's using another service call to effectively counter act those calls if the current script is stopped.

Reply:
I don't understand why we would need direct access to the corresponding entities? 

We're calling with blocking=True, so the cancellation should propagate all the way up (down?) to the script object that is being called.

Reply:
The call to `hass.services.async_call` is indeed with `blocking=True` (and it needs to be that way for other reasons), but the `blocking` parameter of the `script.turn_on` service (which is different) will effectively make it non-blocking overall. So we'll come back from the call to `hass.services.async_call` possibly _before_ the run is done.

To put that another way, for the `script.turn_on` service, when the blocking call to `hass.services.async_call` returns depends on the _service's_ `blocking` _parameter_. If it's `True`, then the call will return when the entire script run has finished (or been stopped.) If it's `False`, then the call will return as soon as the script run _has started_ (and it's state has been updated to `'on'` to indicate it is running.)

Maybe I should change the name of the `script.turn_on` service's parameter to `background` and default it to `False`. Then again, based on the previous discussion, I'll be removing the need to look at the `script.turn_on` service's parameter here, so maybe the confusion will go away anyway.

Reply:
Well I don't think that we should try to stop scripts that we run in the background. If I am in a website and I open a link in another tab, closing the first tab doesn't close the second one. Same with starting threads in Python.

Reply:
Not sure that's the same thing. The idea here is:
```yaml
script:
  a:
    sequence:
    - service: script.turn_on
      entity_id: script.b
      data:
        background: true
    - do_something_else_for_a_while
    - wait_template: "{{ is_state('script.b', 'off') }}"
    - blah_blah_blah
```
Now something starts `script.a`, but then decides that should be stopped so calls `script.turn_off` on it. `script.b` is just an implementation detail of `script.a`. If I want `script.a` to be stopped, then I want _everything it's doing_ to be stopped, including `script.b`.

Probably a better example is:
```yaml
script:
  a:
    mode: restart
    sequence:
    - ...
```
Or even:
```yaml
automation:
- ...
  mode: restart
  action:
  - ...
```

In these cases I want the script to restart for some reason or the automation is triggered again while the actions are still running. That should stop everything it's doing, including any scripts it called directly or indirectly, so it can all start over.

These are just a few simple examples, but it could get more involved with sub-scripts of sub-scripts. I shouldn't have to stop them all individually. That would become a maintenance nightmare.

Reply:
So first, this doesn't really belong in this PR. This seems like a new feature.

Second, since we support simultaneous executions of scripts, it might be that A stopping his own instance of script B by calling turn_off on script B, it is also stopping any other instance of script B that is running in parallel and could have been invoked by other parts of the system.

Reply:
I think it's a natural result of having a restart feature, and I think it does need to be addressed as an inseparable piece of this overall feature set. If a script, or more likely an automation's action sequence, is restarted, it seems natural and expected that anything it does should get stopped so it can all start over again.

Yes, I agree that it is possible for a script to get used by more than one other script and/or automation, and therefore, yes, that would be a side effect -- i.e., one caller gets stopped and all invocations of the "used" script get stopped. I think that would apply equally well whether the script was called in a blocking mode or was started to run in the background. I do see that as a potential problem now that you point it out. But if it's a trade-off between the two behaviors  -- only stopping the outermost script, or stopping all called scripts, recursively, as well -- my opinion is the latter is what would be likely expected and needed. If someone wants a script to run completely independently, they can simply duplicate the script (possibly using YAML anchors.)

I suppose to do it in such a way there are no surprises, whenever one sequence (I mean a Script helper object) invokes another script (and here I mean a script integration entity), maybe it should keep a reference to the task that is executing it (whether in the "foreground" or "background") so it can be canceled (and like you suggested earlier, it can stop itself gracefully.) Maybe that's the best of both worlds.

Reply:
I think that we need a best of both world solutions. Having a script randomly stopped because another script got stopped is not easy to debug. It's a lot easier to debug if something continue to run. It's not possible to keep a reference to the task because we're just calling a service.  I suggest we keep this for a future PR and let's explore if we can use the context ID. But that _will_ get complicated and so we should not have it extend this PR.

Reply:
You're completely right and I totally agree.

I'm not sure how I got on the subject of stopping a background script run. That's not what the code currently does or was intended to do. It is something that needs to be thought about and probably addressed, but that can be done later.

Also, you're right that the stopping of the service call to the script should be done via canceling the task. That should work and highlights the fact that I probably implemented the shielding at the wrong level.

I'll work on this and make another commit along these lines.